### PR TITLE
Speech processing

### PR DESCRIPTION
**Pull request by cifkao**
This PR adds speech processing support to Neural Monkey.

- `audio_reader` reads sound files. It depends on `scipy.io` for reading WAV files and on [`sph2pipe`](https://www.ldc.upenn.edu/language-resources/tools/sphere-conversion-tools) (an external utility) for reading NIST Sphere files.
- `SpeechFeaturesPreprocessor` computes speech features using `python_speech_features`.
- `RawRNNEncoder` is basicaly a copy of `SentenceEncoder` without the embeddings, so that it can be directly fed feature vectors. I cannot think of a nicer way to do this without making drastic changes to `SentenceEncoder`. Maybe the RNN functionality should be moved to a separate class?

There should probably be a test, but I don't know what data to use.

### CODE REVIEW

**Comment by jindrahelcl** [comment]
To je hustý.. :-) Jedinej problém, kterej teď vidím, jsou ty requirements - několikrát zaznělo, že scipy nechceme dávat do requirements, protože je to zbytečně velký. Ale tohle už je třetí případ, kdy ho používáme.

Podle mě by byla škoda, kdybychom contributory nutili nepoužívat scipy. Osobně bych byl pro to, aby se to přidalo do requirements - argumenty proti jsou, že je to moc velký a pomalu by se to instalovalo. Což mi nepřijde jako dostatečná protiváha argumentu, že když to ve scipy je, tak neni důvod dělat to jinak (a složitějc).

Vyvolávám tady o tom diskusi @jlibovicky, @tomasmcz, řekněte, jak si myslíte, že by bylo nejlepší to udělat.

**Comment by tomasmcz** [comment]
Jestli už máme 3 místa, kde je to potřeba a nedá se to jednoduše přepsat do numpy, tak bych to neřešil a dal bych scipy do requirements.

File: neuralmonkey/processors/speech.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+FEATURE_TYPES = {f.__name__: f for f in [mfcc, fbank, logfbank, ssc]}
+
+
+# pylint: disable=invalid-name
+def SpeechFeaturesPreprocessor(feature_type='mfcc', delta_order=2,
+                               delta_window=2, **kwargs) -> Callable:
+    """Calculate speech features.
+
+    By default, compute 13 MFCCs + delta + acceleration, i.e. 39 coefficients.
```
Comment (jlibovicky) [inline neuralmonkey/processors/speech.py]: Možná je to, protože se v téhle oblasti vůbec neorientuju, a ostatním je to zřejmé, ale asi bych jako čtenář dokumentace ocenil, kdyby tu bylo napsáno, že bere ten signál po okýnkách velikosti `delta_window` s posunem `delta_order` a na okýnka uplatní nějakou funkci, ze které vyleze něco, co je najednou ve frekvenční doméně. A jestli nemám pravdu, tak o to víc by to tu mělo být napsáno :-)

File: neuralmonkey/evaluators/wer.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import pyter
+
+
+# pylint: disable=too-few-public-methods
+class WEREvaluator(object):
+    """Compute WER using the pyter library."""
```
Comment (jlibovicky) [inline neuralmonkey/evaluators/wer.py]: Napsal bych celý, že je to Word Error Rate usually used in ASR nebo tak něco.

File: neuralmonkey/encoders/raw_rnn_encoder.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        Arguments:
+            variable: The variable to be dropped out
+
+        Returns:
+            The dropped value of the variable
+        """
+        if self.dropout_keep_p == 1.0:
+            return variable
+
+        # TODO as soon as TF.12 is out, remove this line and use train_mode
```
Comment (jlibovicky) [inline neuralmonkey/encoders/raw_rnn_encoder.py]: @jindrahelcl: Co je tohleto?

File: neuralmonkey/processors/speech.py (hunk 1 above)
Comment (jlibovicky) [inline neuralmonkey/processors/speech.py]: Třeba dát odkaz sem: http://python-speech-features.readthedocs.io/en/latest/

File: neuralmonkey/encoders/raw_rnn_encoder.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+        # directly
+        train_mode_batch = tf.fill(tf.shape(variable)[:1], self.train_mode)
+        dropped_value = tf.nn.dropout(variable, self.dropout_keep_p)
+        return tf.where(train_mode_batch, dropped_value, variable)
+
+    def rnn_cells(self) -> RNNCellTuple:
```
Comment (jlibovicky) [inline neuralmonkey/encoders/raw_rnn_encoder.py]: Nestálo by za to, aby tohle byla base class pro sentence encoder? Jestli se ti to nechce dělat, já to klidně udělám potom, co se tohle zmerguje.

File: neuralmonkey/encoders/raw_rnn_encoder.py (hunk 3 above)
Comment (jindrahelcl) [inline neuralmonkey/encoders/raw_rnn_encoder.py]: tahle funkce by se měla vyhodit a nahradit `nn.utils.dropout`

Ještě jsem nedělal svuj review..

File: neuralmonkey/encoders/raw_rnn_encoder.py (hunk 4 above)
Comment (jindrahelcl) [inline neuralmonkey/encoders/raw_rnn_encoder.py]: nějakej encoder refactoring mam rozdělanej, nechal bych to na potom.

Jedna z věcí, které bychom mohli napsat, je embedding encoder, kterej by sekvenci indexů do slovníku převáděl na embeddingy. A pak ještě udělat nějakej trik, kterým by se pak do parametru `data_id` resp. `data_ids` daly narvat enkodéry.. To by ale bylo asi už příliš ezoterický. Na začátek bych se omezil na abstraktní RNN encoder, kterej by měl syny.

**Comment by jlibovicky** [comment]
Scipy asi klidně můžeme přidat do requirements. Možná bych pak mohl přepsat načítání obrázků pomocí `pillow` do scipy, abychom měli o depedency méně.

File: neuralmonkey/encoders/raw_rnn_encoder.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        self.inputs = tf.placeholder(tf.float32,
+                                     shape=[None, None,
+                                            self.input_dimension],
+                                     name="encoder_input")
+
+        self._input_lengths = tf.placeholder(
+            tf.int32, shape=[None],
+            name="encoder_padding_lengths")
+
+    def _dropout(self, variable: tf.Tensor) -> tf.Tensor:
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/raw_rnn_encoder.py]: jak už tu je napsaný - nahradit funkcí `neuralmonkey.utils.dropout`

File: neuralmonkey/encoders/raw_rnn_encoder.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    def __init__(self,
+                 name: str,
+                 data_id: str,
+                 rnn_size: int,
+                 input_dimension: int,
+                 max_input_len: Optional[int]=None,
+                 dropout_keep_prob: float=1.0,
+                 attention_type: Optional[Any]=None,
+                 attention_fertility: int=3,
+                 use_noisy_activations: bool=False,
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/raw_rnn_encoder.py]: vim, že je to kopie sentence enkoderu, ale tyhle věci tu být nemusí:
`attention_fertility`, `use_noisy_activations` a `parent_encoder`.. To tady trochu zjednoduší ten kód.

Comment (jindrahelcl) [inline neuralmonkey/encoders/raw_rnn_encoder.py]: (o tý fertility bychom mohli vést diskusi, co by to asi tak dělalo se speech signálem)

File: neuralmonkey/encoders/raw_rnn_encoder.py [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+    @property
+    def _attention_tensor(self):
+        return self.__attention_tensor
+
+    @property
+    def _attention_mask(self):
+        return self._input_mask
+
+    def _create_input_placeholders(self):
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/raw_rnn_encoder.py]: že vrací None tady chybí

File: neuralmonkey/encoders/raw_rnn_encoder.py [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+            self.encoded = tf.concat(encoded_tup, 1)
+
+        log("RNN encoder initialized")
+
+    @property
+    def _attention_tensor(self):
+        return self.__attention_tensor
+
+    @property
+    def _attention_mask(self):
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/raw_rnn_encoder.py]: tady u těch dvou properties by měla bejt anotace, že vracej tf.Tensor

File: neuralmonkey/evaluators/wer.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
+import pyter
+
+
+# pylint: disable=too-few-public-methods
+class WEREvaluator(object):
```
Comment (jindrahelcl) [inline neuralmonkey/evaluators/wer.py]: přidej prosím sem do tý třídy type hinty

File: neuralmonkey/processors/speech.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
+from typing import Callable
+
+import numpy as np
+from python_speech_features import mfcc, fbank, logfbank, ssc, delta
+
+from neuralmonkey.readers.audio_reader import Audio
+
+
+# pylint: disable=invalid-name
+def SpeechFeaturesPreprocessor(feature_type='mfcc', delta_order=0,
```
Comment (jindrahelcl) [inline neuralmonkey/processors/speech.py]: tohle taky kdyžtak type hinty u argumentů - jsou tam potřeba, i když je daná defaultní hodnota, tj, třeba
`feature_type: str='mfcc'`.. U `*args` a `**kwargs` podle mě být nemusí.

**Comment by cifkao** [comment]
Nevím, co si mám myslet o tom [segfaultu v travis-ci/pr](https://travis-ci.org/ufal/neuralmonkey/jobs/214303184#L2689). Spadlo to v audio-classifier.ini asi někde mezi načtením datasetu a inicializací session. travis-ci/push je OK. (Ale po rebasu, který jsem teď udělal, by v nich asi neměl být rozdíl?)

**Comment by cifkao** [comment]
…a teď je najednou pod tím samým odkazem běh, který je zelený. Rozumíte tomu?

**Comment by jlibovicky** [comment]
Netuším, když jsem to pustil na Travisu ručně znovu, tak to prošlo. Bug asi nebude na naší straně.
